### PR TITLE
Refactor `AllowIncompatibleOverride`

### PR DESCRIPTION
**Pull request by sambostock**
This makes the following changes:

- Fixes a typo in the cop description.
- Switches from the deprecated `Cop` parent class to `Base`
- Removes the unneeded `not_nil?` matcher
- Switches from a node search to order independent matching for the
  `allow_incompatible` `pair`
- Uses the idiomatic `MSG` constant
- Uses `RESTRICT_ON_SEND` to optimize the `on_send` calls, avoiding
  irrelevant method calls
- Uses a capture in the node matcher instead of indexing into children
- Adds tests for edge cases

### CODE REVIEW

**Review by vinistock — APPROVED** [review]

**Review by paracycle — APPROVED** [review]
Just one small formatting suggestion.

File: spec/rubocop/cop/sorbet/signatures/allow_incompatible_override_spec.rb [hunk 1]
```diff
@@ -3,24 +3,39 @@
...
-    "Usage of `allow_incompatible` suggests a violation of the Liskov Substitution Principle. "\
-    "Instead, strive to write interfaces which respect subtyping principles and remove `allow_incompatible`"
-  end
+  let(:message) { RuboCop::Cop::Sorbet::AllowIncompatibleOverride::MSG }
 
   it("disallows using override(allow_incompatible: true)") do
     expect_offense(<<~RUBY)
       class Foo
-        sig(a: Integer).override(allow_incompatible: true).void
+        sig(a: Integer).override(allow_incompatible: true, something: :unrelated).void
```
Comment (sambostock) [inline spec/rubocop/cop/sorbet/signatures/allow_incompatible_override_spec.rb:11]: This test shouldn't have extra keywords:

```suggestion
        sig(a: Integer).override(allow_incompatible: true).void
```

I added the test below to test that separately

File: lib/rubocop/cop/sorbet/signatures/allow_incompatible_override.rb [hunk 2]
```diff
@@ -18,43 +18,28 @@ module Sorbet
...
-          (
-            send
-            [#not_nil? #sig?]
+        def_node_matcher(:allow_incompatible_override?, <<~PATTERN)
+          (send
+            #sig?
             :override
-            [#not_nil? #allow_incompatible?]
-          )
+            (hash <$(pair (sym :allow_incompatible) true) ...>))
```
[the remark this answers is not in the document]
Comment (sambostock) [reply lib/rubocop/cop/sorbet/signatures/allow_incompatible_override.rb:31]: Yeah, fair. I bounce around between parens on their own line, or parens on the same line, since the `ruby-parse` output spits them out on the same line, and I often copy paste that as the starting point for a node pattern.

**Review by sambostock — COMMENTED** [review]
I've just realized this cop is looking for the old syntax, whereas we don't write sigs like this anymore

```diff
-sig(a: Integer).override.void
+sig { override.params(a: Integer).void }
```

We can tackle supporting that in a separate PR though.
